make detector library required

We currently have some code supporting building `checkpoint` (and `vt`) without `detector` lib. If we are confident that we don't need to support such use case, we can remove it.

```bash
vt/lib/checkpoint > git grep HAS_DETECTION_COMPONENT

examples/checkpoint_example_3.cc:126:#if HAS_DETECTION_COMPONENT
examples/checkpoint_example_3.cc:126:#if HAS_DETECTION_COMPONENT
src/CMakeLists.txt:80:target_compile_definitions(${CHECKPOINT_LIBRARY} PUBLIC HAS_DETECTION_COMPONENT=1)
src/checkpoint/dispatch/dispatch_serializer_byte.h:58:#if !HAS_DETECTION_COMPONENT
src/checkpoint/dispatch/dispatch_serializer_byte.h:73:  #if HAS_DETECTION_COMPONENT
src/checkpoint/dispatch/dispatch_serializer_nonbyte.h:63:#if !HAS_DETECTION_COMPONENT
src/checkpoint/dispatch/dispatch_serializer_nonbyte.h:106:  #if HAS_DETECTION_COMPONENT
src/checkpoint/dispatch/reconstructor.h:70:  #if HAS_DETECTION_COMPONENT
src/checkpoint/dispatch/reconstructor.h:121:    #if HAS_DETECTION_COMPONENT
src/checkpoint/dispatch/reconstructor.h:140:  #if HAS_DETECTION_COMPONENT
src/checkpoint/dispatch/vrt/virtual_serialize_traits.h:53:#if HAS_DETECTION_COMPONENT
src/checkpoint/dispatch/vrt/virtual_serialize_traits.h:55:#endif /*HAS_DETECTION_COMPONENT*/
src/checkpoint/dispatch/vrt/virtual_serialize_traits.h:59:#if HAS_DETECTION_COMPONENT
src/checkpoint/traits/checkpoint_traits.h:52:#if HAS_DETECTION_COMPONENT
src/checkpoint/traits/checkpoint_traits.h:54:#endif  /*HAS_DETECTION_COMPONENT*/
src/checkpoint/traits/checkpoint_traits.h:56:#if HAS_DETECTION_COMPONENT
src/checkpoint/traits/checkpoint_traits.h:100:#endif  /*HAS_DETECTION_COMPONENT*/
src/checkpoint/traits/serializable_traits.h:55:#if HAS_DETECTION_COMPONENT
src/checkpoint/traits/serializable_traits.h:57:#endif  /*HAS_DETECTION_COMPONENT*/
src/checkpoint/traits/serializable_traits.h:59:#if HAS_DETECTION_COMPONENT
src/checkpoint/traits/serializable_traits.h:220:#endif  /*HAS_DETECTION_COMPONENT*/
```

- [x] try to build without `detector` to see if that's even possible
- [x] confirm that `detector` can be set as required
- [x] remove obsolete code and update docs accordingly
- [x] make sure that vt builds correctly after the changes

Note: [`vt_detector_disabled`](https://github.com/DARMA-tasking/vt/blob/2f2e7b4b8f1db5dec6c2687858e273b460a90571/cmake/define_build_types.cmake#L22) option is not tested in CI and `vt` fails to build when it is enabled:
```
/bin/clang++-10  -DFMT_HEADER_ONLY=1 -DFMT_USE_USER_DEFINED_LITERALS=0 -DHAS_DETECTION_COMPONENT=1 -I../lib/fmt -I../lib/CLI -Idebug -I../src -I../lib/libfort/lib -Ilib/checkpoint/src -I../lib/checkpoint/src -I../lib/detector/src -isystem /usr/include/x86_64-linux-gnu/mpich   -Wall -pedantic -Wshadow -Wno-unknown-pragmas -Wsign-compare -ftemplate-backtrace-limit=100 -ftemplate-depth=900 -DCLI11_EXPERIMENTAL_OPTIONAL=0 -Werror -g   -fcolor-diagnostics -fPIC -std=c++14 -o src/CMakeFiles/vt.dir/vt/context/runnable_context/lb_stats.cc.o -c /home/cz4rs/nga/vt/src/vt/context/runnable_context/lb_stats.cc
In file included from /home/cz4rs/nga/vt/src/vt/context/runnable_context/lb_stats.cc:46:
In file included from ../src/vt/vrt/collection/manager.h:50:
In file included from ../src/vt/vrt/collection/manager.fwd.h:49:
In file included from ../src/vt/vrt/collection/dispatch/dispatch.h:49:
../src/vt/vrt/collection/traits/coll_msg.h:60:32: error: use of undeclared identifier 'detection'
  using hasCollectionMsgType = detection::is_detected<IsCollectionMsgType, T>;
                               ^
../src/vt/vrt/collection/traits/coll_msg.h:60:32: error: expected a type
../src/vt/vrt/collection/traits/coll_msg.h:60:43: error: expected ';' after alias declaration
  using hasCollectionMsgType = detection::is_detected<IsCollectionMsgType, T>;
                                          ^
                                          ;
../src/vt/vrt/collection/traits/coll_msg.h:61:45: error: use of undeclared identifier 'hasCollectionMsgType'
  static constexpr auto const is_coll_msg = hasCollectionMsgType::value;
                                            ^
```

--- a/docs/md/checkpoint_learn_build.md
+++ b/docs/md/checkpoint_learn_build.md
@@ -6,9 +6,12 @@ checkpoint can be built with `cmake`.
 
 To build *checkpoint*, one must obtain the following dependencies:
 
-\subsection ckpt_optional-deps Optional
+\subsection ckpt_required-deps Required
 
   - [detector](https://github.com/DARMA-tasking/detector)
+
+\subsection ckpt_optional-deps Optional
+
   - gtest,   (if testing *checkpoint* is enabled)
   - MPI,   (if MPI tests are enabled)
   - [Kokkos](https://github.com/kokkos/kokkos),  (if Kokkos views need to be serialized)
@@ -16,9 +19,9 @@ To build *checkpoint*, one must obtain the following dependencies:
 
 \subsection ckpt_use-cmake-directly-vars Using cmake directly
 
-One may use `cmake` as normal on *checkpoint* once the dependencies are
-installed.  The following are some configuration build options that can be
-provided to `cmake` to change the build configuration:
+One may use `cmake` to build *checkpoint* once the dependencies are installed.
+The following are some options that can be provided to `cmake` to change the
+build configuration:
 
 | CMake Variable                    | Default Value   | Description                           |
 | ------------------                | --------------- | -----------                           |

--- a/examples/checkpoint_example_3.cc
+++ b/examples/checkpoint_example_3.cc
@@ -123,33 +123,32 @@ struct TestReconstruct {
 
 }} // end namespace checkpoint::examples
 
-#if HAS_DETECTION_COMPONENT
-  #include "checkpoint/traits/serializable_traits.h"
+#include "checkpoint/traits/serializable_traits.h"
 
-  namespace checkpoint {
+namespace checkpoint {
 
-  using namespace examples;
+using namespace examples;
 
-  static_assert(
-    SerializableTraits<TestDefaultCons>::is_serializable,
-    "Should be serializable"
-  );
-  static_assert(
-    ! SerializableTraits<TestNoSerialize>::is_serializable,
-    "Should not be serializable"
-  );
+static_assert(
+  SerializableTraits<TestDefaultCons>::is_serializable,
+  "Should be serializable"
+);
+static_assert(
+  ! SerializableTraits<TestNoSerialize>::is_serializable,
+  "Should not be serializable"
+);
 
-  static_assert(
-    ! SerializableTraits<TestShouldFailReconstruct>::is_serializable,
-    "Should not be serializable"
-  );
+static_assert(
+  ! SerializableTraits<TestShouldFailReconstruct>::is_serializable,
+  "Should not be serializable"
+);
 
-  static_assert(
-    SerializableTraits<TestReconstruct>::is_serializable,
-    "Should be serializable"
-  );
-  } // end namespace checkpoint
-#endif
+static_assert(
+  SerializableTraits<TestReconstruct>::is_serializable,
+  "Should be serializable"
+);
+} // end namespace checkpoint
+
 
 int main(int, char**) {
   // Example is a compile-time test of serializability traits

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -77,8 +77,6 @@ add_library(${CHECKPOINT_LIBRARY_NS} ALIAS ${CHECKPOINT_LIBRARY})
 target_compile_options(${CHECKPOINT_LIBRARY} PUBLIC "-fPIC")
 target_compile_options(${CHECKPOINT_LIBRARY} PUBLIC ${CXX_STANDARD_FLAGS})
 
-target_compile_definitions(${CHECKPOINT_LIBRARY} PUBLIC HAS_DETECTION_COMPONENT=1)
-
 include(CMakePrintHelpers)
 
 if (${kokkos_DIR_FOUND})

--- a/src/checkpoint/dispatch/dispatch_serializer_byte.h
+++ b/src/checkpoint/dispatch/dispatch_serializer_byte.h
@@ -55,42 +55,15 @@
 
 namespace checkpoint { namespace dispatch {
 
-#if !HAS_DETECTION_COMPONENT
-template <typename T>
-struct hasByteCopy {
-  template <typename C, typename = typename C::isByteCopyable>
-  static typename C::isByteCopyable test(int);
-
-  template <typename C>
-  static std::false_type test(...);
-
-  static constexpr bool value = decltype(test<T>(0))::value;
-};
-#endif
-
 template <typename SerializerT, typename T, typename Dispatcher>
 struct SerializerDispatchByte {
-  #if HAS_DETECTION_COMPONENT
-    template <typename U>
-    using isByteCopyType =
-    typename std::enable_if<SerializableTraits<U,void>::is_bytecopyable, T>::type;
+  template <typename U>
+  using isByteCopyType =
+  typename std::enable_if<SerializableTraits<U,void>::is_bytecopyable, T>::type;
 
-    template <typename U>
-    using isNotByteCopyType =
-    typename std::enable_if<not SerializableTraits<U,void>::is_bytecopyable, T>::type;
-  #else
-    template <typename U>
-    using isByteCopyType =
-    typename std::enable_if<
-      std::is_arithmetic<U>::value or hasByteCopy<U>::value, T
-    >::type;
-
-    template <typename U>
-    using isNotByteCopyType =
-    typename std::enable_if<
-      not std::is_arithmetic<U>::value and not hasByteCopy<U>::value, T
-    >::type;
-  #endif
+  template <typename U>
+  using isNotByteCopyType =
+  typename std::enable_if<not SerializableTraits<U,void>::is_bytecopyable, T>::type;
 
   template <typename U = T>
   void operator()(

--- a/src/checkpoint/dispatch/dispatch_serializer_nonbyte.h
+++ b/src/checkpoint/dispatch/dispatch_serializer_nonbyte.h
@@ -57,25 +57,6 @@
 
 namespace checkpoint { namespace dispatch {
 
-// Implement a lightweight simple detection mechanism (when full detector is not
-// present) to detect type of serialize() method: intrusive/non-intrusive
-// variants
-#if !HAS_DETECTION_COMPONENT
-template <typename SerializerT, typename T>
-struct hasSerialize {
-  template <
-    typename C,
-    typename = decltype(std::declval<C>().serialize(std::declval<SerializerT&>()))
-  >
-  static std::true_type test(int);
-
-  template <typename C>
-  static std::false_type test(...);
-
-  static constexpr bool value = decltype(test<T>(0))::value;
-};
-#endif
-
 template <typename SerializerT, typename T>
 struct BasicDispatcher {
   static void serializeIntrusive(SerializerT& s, T& t) {
@@ -101,54 +82,42 @@ struct SerializerDispatchNonByte {
     apply(s, val, num);
   }
 
-  // If we have the detection component, we can more precisely check for
-  // serializability
-  #if HAS_DETECTION_COMPONENT
-    template <typename U>
-    using hasSplitSerialize =
-    typename std::enable_if<SerializableTraits<U,S>::has_split_serialize, T>::type;
+  template <typename U>
+  using hasSplitSerialize =
+  typename std::enable_if<SerializableTraits<U,S>::has_split_serialize, T>::type;
 
-    template <typename U>
-    using hasNotSplitSerialize =
-    typename std::enable_if<!SerializableTraits<U,S>::has_split_serialize && !std::is_enum<U>::value, T>::type;
+  template <typename U>
+  using hasNotSplitSerialize =
+  typename std::enable_if<!SerializableTraits<U,S>::has_split_serialize && !std::is_enum<U>::value, T>::type;
 
-    template <typename U>
-    using hasInSerialize =
-    typename std::enable_if<SerializableTraits<U,S>::has_serialize_instrusive, T>::type;
+  template <typename U>
+  using hasInSerialize =
+  typename std::enable_if<SerializableTraits<U,S>::has_serialize_instrusive, T>::type;
 
-    template <typename U>
-    using hasNoninSerialize =
-    typename std::enable_if<SerializableTraits<U,S>::has_serialize_noninstrusive, T>::type;
+  template <typename U>
+  using hasNoninSerialize =
+  typename std::enable_if<SerializableTraits<U,S>::has_serialize_noninstrusive, T>::type;
 
-    template <typename U>
-    using hasVirtualSerialize =
-    typename std::enable_if<vrt::VirtualSerializeTraits<U>::has_virtual_serialize, T>::type;
+  template <typename U>
+  using hasVirtualSerialize =
+  typename std::enable_if<vrt::VirtualSerializeTraits<U>::has_virtual_serialize, T>::type;
 
-    template <typename U>
-    using hasNotVirtualSerialize =
-    typename std::enable_if<vrt::VirtualSerializeTraits<U>::has_not_virtual_serialize, T>::type;
+  template <typename U>
+  using hasNotVirtualSerialize =
+  typename std::enable_if<vrt::VirtualSerializeTraits<U>::has_not_virtual_serialize, T>::type;
 
-    template <typename U>
-    using isEnum =
-    typename std::enable_if<std::is_enum<U>::value, T>::type;
+  template <typename U>
+  using isEnum =
+  typename std::enable_if<std::is_enum<U>::value, T>::type;
 
-    template <typename U>
-    using justFootprint =
-    typename std::enable_if<
-      std::is_same<S, checkpoint::Footprinter>::value and
-      not SerializableTraits<U, S>::is_traversable and
-      not std::is_enum<U>::value,
-      T
-    >::type;
-  #else
-    template <typename U>
-    using hasInSerialize =
-    typename std::enable_if<hasSerialize<SerializerT, U>::value, T>::type;
-
-    template <typename U>
-    using hasNoninSerialize =
-    typename std::enable_if<!hasSerialize<SerializerT, U>::value, T>::type;
-  #endif
+  template <typename U>
+  using justFootprint =
+  typename std::enable_if<
+    std::is_same<S, checkpoint::Footprinter>::value and
+    not SerializableTraits<U, S>::is_traversable and
+    not std::is_enum<U>::value,
+    T
+  >::type;
 
   template <typename U = T>
   void applyElm(

--- a/src/checkpoint/dispatch/reconstructor.h
+++ b/src/checkpoint/dispatch/reconstructor.h
@@ -65,47 +65,37 @@ struct Reconstructor {
   using isNotDefaultConsType =
   typename std::enable_if<not std::is_default_constructible<U>::value, T>::type;
 
-  // If we have the detection component, we can more precisely check for
-  // reconstuctibility
-  #if HAS_DETECTION_COMPONENT
-    template <typename U>
-    using isReconstructibleType =
-    typename std::enable_if<
-      SerializableTraits<U,void>::is_intrusive_reconstructible, T
-    >::type;
+  template <typename U>
+  using isReconstructibleType =
+  typename std::enable_if<
+    SerializableTraits<U,void>::is_intrusive_reconstructible, T
+  >::type;
 
-    template <typename U>
-    using isNonIntReconstructibleType =
-    typename std::enable_if<
-      SerializableTraits<U,void>::is_nonintrusive_reconstructible, T
-    >::type;
+  template <typename U>
+  using isNonIntReconstructibleType =
+  typename std::enable_if<
+    SerializableTraits<U,void>::is_nonintrusive_reconstructible, T
+  >::type;
 
-    template <typename U>
-    using isNotReconstructibleType =
-    typename std::enable_if<not SerializableTraits<U,void>::is_reconstructible, T>::type;
+  template <typename U>
+  using isNotReconstructibleType =
+  typename std::enable_if<not SerializableTraits<U,void>::is_reconstructible, T>::type;
 
-    template <typename U>
-    using isTaggedConstructibleType =
-    typename std::enable_if<SerializableTraits<U,void>::is_tagged_constructible, T>::type;
+  template <typename U>
+  using isTaggedConstructibleType =
+  typename std::enable_if<SerializableTraits<U,void>::is_tagged_constructible, T>::type;
 
-    template <typename U>
-    using isNotTaggedConstructibleType =
-    typename std::enable_if<not SerializableTraits<U,void>::is_tagged_constructible, T>::type;
+  template <typename U>
+  using isNotTaggedConstructibleType =
+  typename std::enable_if<not SerializableTraits<U,void>::is_tagged_constructible, T>::type;
 
-    template <typename U>
-    using isConstructible =
-    typename std::enable_if<SerializableTraits<U,void>::is_constructible, T>::type;
+  template <typename U>
+  using isConstructible =
+  typename std::enable_if<SerializableTraits<U,void>::is_constructible, T>::type;
 
-    template <typename U>
-    using isNotConstructible =
-    typename std::enable_if<not SerializableTraits<U,void>::is_constructible, T>::type;
-  #else
-    template <typename U>
-    using isConstructible = isDefaultConsType<U>;
-
-    template <typename U>
-    using isNotConstructible = isNotDefaultConsType<U>;
-  #endif
+  template <typename U>
+  using isNotConstructible =
+  typename std::enable_if<not SerializableTraits<U,void>::is_constructible, T>::type;
 
   // Default-construct as lowest priority in reconstruction preference
   template <typename U = T>
@@ -118,7 +108,6 @@ struct Reconstructor {
   // Fail, no valid option to constructing T
   template <typename U = T>
   static T* constructDefault(void* buf, isNotDefaultConsType<U>* = nullptr) {
-    #if HAS_DETECTION_COMPONENT
     static_assert(
       SerializableTraits<U,void>::is_tagged_constructible or
       SerializableTraits<U,void>::is_reconstructible or
@@ -126,18 +115,8 @@ struct Reconstructor {
       "Either a default constructor, reconstruct() function, or tagged "
       "constructor are required for de-serialization"
     );
-    #else
-    static_assert(
-      std::is_default_constructible<U>::value,
-      "A default constructor is required for de-serialization. To enable "
-      "reconstruct or tagged constructors, you must compile with the "
-      "detection component."
-    );
-    #endif
     return nullptr;
   }
-
-  #if HAS_DETECTION_COMPONENT
 
   /*
    * Try to reconstruct with the following precedence:
@@ -191,15 +170,6 @@ struct Reconstructor {
   static T* construct(void* buf) {
     return constructTag<U>(buf);
   }
-
-  #else
-
-  template <typename U = T>
-  static T* construct(void* buf) {
-    return constructDefault<U>(buf);
-  }
-
-  #endif
 
   /// Overloads that allow failure to reconstruct so SFINAE overloads don't
   /// static assert out

--- a/src/checkpoint/dispatch/vrt/virtual_serialize_traits.h
+++ b/src/checkpoint/dispatch/vrt/virtual_serialize_traits.h
@@ -48,15 +48,11 @@
 #include "checkpoint/common.h"
 #include "checkpoint/dispatch/vrt/registry_common.h"
 
+#include "detector_headers.h"
+
 #include <type_traits>
 
-#if HAS_DETECTION_COMPONENT
-#include "detector_headers.h"
-#endif /*HAS_DETECTION_COMPONENT*/
-
 namespace checkpoint { namespace dispatch { namespace vrt {
-
-#if HAS_DETECTION_COMPONENT
 
 template <typename T>
 struct VirtualSerializeTraits {
@@ -107,30 +103,6 @@ struct VirtualSerializeTraits {
   static constexpr auto const has_not_virtual_serialize =
     not has_virtual_serialize;
 };
-
-#else
-
-template <typename T>
-struct VirtualSerializeTraits {
-  template <
-    typename C,
-    typename = decltype(
-      std::declval<U>()._checkpointDynamicSerialize(
-        std::declval<void*>(),
-        std::declval<TypeIdx>(),
-        std::declval<TypeIdx>()
-      )
-    )
-  >
-  static std::true_type test(int);
-
-  template <typename C>
-  static std::false_type test(...);
-
-  static constexpr bool is_virtual_serializable = decltype(test<T>(0))::value;
-};
-
-#endif
 
 }}} /* end namespace checkpoint::dispatch::vrt */
 

--- a/src/checkpoint/traits/checkpoint_traits.h
+++ b/src/checkpoint/traits/checkpoint_traits.h
@@ -49,11 +49,7 @@
 
 #include "checkpoint/common.h"
 
-#if HAS_DETECTION_COMPONENT
 #include "detector_headers.h"
-#endif  /*HAS_DETECTION_COMPONENT*/
-
-#if HAS_DETECTION_COMPONENT
 
 namespace checkpoint {
 
@@ -96,7 +92,5 @@ struct SerializerTraits {
 };
 
 }  // end namespace checkpoint
-
-#endif  /*HAS_DETECTION_COMPONENT*/
 
 #endif /*INCLUDED_CHECKPOINT_TRAITS_CHECKPOINT_TRAITS_H*/

--- a/src/checkpoint/traits/serializable_traits.h
+++ b/src/checkpoint/traits/serializable_traits.h
@@ -49,14 +49,10 @@
 #include "checkpoint/serializers/serializers_headers.h"
 #include "checkpoint/dispatch/reconstructor_tag.h"
 
+#include "detector_headers.h"
+
 #include <cstdint>
 #include <cassert>
-
-#if HAS_DETECTION_COMPONENT
-#include "detector_headers.h"
-#endif  /*HAS_DETECTION_COMPONENT*/
-
-#if HAS_DETECTION_COMPONENT
 
 /*
  * Start the traits class with all the archetypes that can be detected for
@@ -216,7 +212,5 @@ struct SerializableTraits {
 };
 
 }  // end namespace checkpoint
-
-#endif  /*HAS_DETECTION_COMPONENT*/
 
 #endif /*INCLUDED_CHECKPOINT_TRAITS_SERIALIZABLE_TRAITS_H*/

